File: module-6-context/solutions/python/compaction.py
"""Module 6: Context compaction."""
import json
import anthropic

client = anthropic.Anthropic()


def estimate_tokens(messages: list) -> int:
    return len(json.dumps(messages, default=str)) // 4


def compact_history(messages: list, token_limit: int = 80000) -> list:
    estimated = estimate_tokens(messages)
    if estimated < token_limit:
        return messages

    first = messages[0]
    recent = messages[-6:]
    middle = messages[1:-6]

    if not middle:
        return messages

    print(f"[Compaction] {estimated} estimated tokens > {token_limit} limit")
    print(f"[Compaction] Summarizing {len(middle)} middle messages...")

    response = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=1024,
        messages=[{
            "role": "user",
            "content": f"Summarize the following agent conversation history concisely. Focus on: what was accomplished, key decisions made, files modified, and current state.\n\n{json.dumps(middle, default=str)}",
        }],
    )

    summary_text = response.content[0].text if response.content[0].type == "text" else "(summary failed)"
    print(f"[Compaction] Reduced to summary ({len(summary_text)} chars)")

    return [first, {"role": "assistant", "content": f"[Previous conversation summary]\n{summary_text}"}, *recent]

File: module-6-context/solutions/compaction.ts
/**
 * Module 6: Context compaction
 * Summarize old turns when approaching token limit.
 */
import Anthropic from "@anthropic-ai/sdk";
import type { Message } from "../../module-2-agent-loop/solutions/types.js";

const client = new Anthropic();

// Rough token estimate: ~4 chars per token
function estimateTokens(messages: Message[]): number {
  return Math.ceil(JSON.stringify(messages).length / 4);
}

export async function compactHistory(
  messages: Message[],
  tokenLimit: number = 80000
): Promise<Message[]> {
  const estimated = estimateTokens(messages);
  if (estimated < tokenLimit) return messages;

  // Keep the first user message and the last 6 messages
  const first = messages[0];
  const recent = messages.slice(-6);
  const middle = messages.slice(1, -6);

  if (middle.length === 0) return messages;

  console.log(`[Compaction] ${estimated} estimated tokens > ${tokenLimit} limit`);
  console.log(`[Compaction] Summarizing ${middle.length} middle messages...`);

  const summaryResponse = await client.messages.create({
    model: "claude-sonnet-4-20250514",
    max_tokens: 1024,
    messages: [
      {
        role: "user",
        content: `Summarize the following agent conversation history concisely. Focus on: what was accomplished, key decisions made, files modified, and current state.\n\n${JSON.stringify(middle)}`,
      },
    ],
  });

  const summaryText =
    summaryResponse.content[0].type === "text"
      ? summaryResponse.content[0].text
      : "(summary failed)";

  console.log(`[Compaction] Reduced to summary (${summaryText.length} chars)`);

  return [
    first,
    { role: "assistant", content: `[Previous conversation summary]\n${summaryText}` },
    ...recent,
  ];
}

export { estimateTokens };
